### PR TITLE
Fix error in multipart upload when merger is empty for skewed datasets

### PR DESCRIPTION
**Pull request by derekluan**
In a skewed distribution, some calls to s3_utils.multipart_upload will not result in any chunks being uploaded to S3.

complete_multipart_upload will throw an xml error if it is called and no chunks were uploaded. Therefore we should only call it if we have uploaded chunks.

### CODE REVIEW

File: raysort/s3_utils.py [hunk 1]
```diff
@@ -176,11 +176,13 @@ def _upload_part(data):
         mpu_id, resp = mpu_queue.get()
         mpu_parts.append({"ETag": resp["ETag"], "PartNumber": mpu_id})
 
-    s3_client.complete_multipart_upload(
-        Bucket=pinfo.bucket,
-        Key=pinfo.path,
-        MultipartUpload={"Parts": mpu_parts},
-        UploadId=mpu["UploadId"],
-    )
+    # only complete the multipart upload if data was uploaded
```
Comment (franklsf95) [inline raysort/s3_utils.py:179]: No need to comment.

File: raysort/s3_utils.py [hunk 2]
```diff
@@ -176,11 +176,13 @@ def _upload_part(data):
...
+    if len(mpu_parts) > 0:
+        s3_client.complete_multipart_upload(
+            Bucket=pinfo.bucket,
+            Key=pinfo.path,
+            MultipartUpload={"Parts": mpu_parts},
+            UploadId=mpu["UploadId"],
+        )
     pinfo.size = bytes_count
     return [pinfo]
```
Comment (franklsf95) [inline raysort/s3_utils.py:188]: As we discussed, return `[]` if no data was uploaded (and hence the object won't exist).

File: scripts/config/terraform/aws-template/main.tf [hunk 3]
```diff
@@ -21,7 +21,7 @@ resource "aws_instance" "cluster" {
   associate_public_ip_address = true
   # subnet_id                   = "subnet-084e54bf496121333" # exoshuffle-subnet-public1-us-west-2a
   # subnet_id                   = "subnet-0af4497f8b2bffeaa" # exoshuffle-subnet-public1-us-west-2b
-  subnet_id = "subnet-033a4bd05bcc8c21e" # exoshuffle-subnet-public1-us-west-2c
+  # subnet_id = "subnet-033a4bd05bcc8c21e" # exoshuffle-subnet-public1-us-west-2c
```
Comment (franklsf95) [inline scripts/config/terraform/aws-template/main.tf:24]: Uncommit this please.

**Review by franklsf95 — COMMENTED** [review]
Thanks for the fix! Nit on PR titles: it should be a (somewhat detailed) description of what is done. So I'd rename it "Fix error in multipart upload when merger is empty for skewed datasets" (also capitalize the first letter).

File: raysort/s3_utils.py (hunk 2 above)
Comment (derekluan) [reply raysort/s3_utils.py:188]: got it, haven't had a chance to test it yet

File: raysort/s3_utils.py [hunk 4]
```diff
@@ -176,11 +176,14 @@ def _upload_part(data):
...
+    if len(mpu_parts) > 0:
+        s3_client.complete_multipart_upload(
+            Bucket=pinfo.bucket,
+            Key=pinfo.path,
+            MultipartUpload={"Parts": mpu_parts},
+            UploadId=mpu["UploadId"],
+        )
+        pinfo.size = bytes_count
+        return [pinfo]
+    else:
```
Comment (franklsf95) [inline raysort/s3_utils.py:188]: `else` is redundant. Just `return [None]`

Can you also change this function's type signature to `list[Optional[PartInfo]]`? This also means `main_push.py:final_merge()`'s return value is now `Optional[PartInfo]`.

**Review by franklsf95 — APPROVED** [review]
See inline comments. Then it's good to merge.

**Pull request merged**